### PR TITLE
gh-1646 Token amount component

### PR DESCRIPTION
**Pull request by elgatovital**
Handles #1646

Changes proposed in this pull request:
- Add component for displaying & editing the token amount

### CODE REVIEW

File: Multisig/UI/UI Library/Views/TokenAmountField.swift [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    @IBOutlet weak var amountTextField: UITextField!
+    @IBOutlet weak var symbolLabel: UILabel!
+    @IBOutlet weak var errorLabel: UILabel!
+    
+    var balance: String {
+        get { amountTextField.text ?? "" }
+        set { amountTextField.text = newValue }
+    }
+    
+    override func commonInit() {
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/Views/TokenAmountField.swift]: you can remove this as it only calls `super`

File: Multisig/UI/UI Library/Views/TokenAmountField.swift [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        super.commonInit()
+    }
+    
+    override func awakeFromNib() {
+        symbolLabel.setStyle(.tertiary)
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/Views/TokenAmountField.swift:26]: something is missing: you must to call super.awakeFromNib() in this method

File: Multisig/UI/UI Library/Views/TokenAmountField.swift [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        errorLabel.isHidden = false
+        borderImage.tintColor = .error
+    }
+}
+
+// allow only decimal numbers to be entered in the amount field
+extension TokenAmountField: UITextFieldDelegate {
+    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
+        
+        let decimalSeparator = Locale.autoupdatingCurrent.decimalSeparator ?? "."
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/Views/TokenAmountField.swift]: decimal separator can be 1 of 3 different things (this is how keyboard works), please see more here: https://github.com/gnosis/SwiftCryptoTokenFormatter/blob/6f605fd7bd61327a6006919e28bf3f0632d1c556/SwiftCryptoTokenFormatter/SwiftCryptoTokenFormatter/TokenFormatter.swift#L14

File: Multisig/UI/UI Library/Views/TokenAmountField.swift [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        let inverseSet = CharacterSet(charactersIn:"0123456789").inverted
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/Views/TokenAmountField.swift]: there's a `decimalDigit` set already: https://developer.apple.com/documentation/foundation/characterset/1779641-decimaldigits#

File: Multisig/UI/UI Library/Views/TokenAmountField.swift (hunk 3 above)
Comment (DmitryBespalov) [reply Multisig/UI/UI Library/Views/TokenAmountField.swift]: and it doesn't adhere to the locale. I killed a lot of time in the past to find out this.

File: Multisig/UI/UI Library/Views/TokenAmountField.swift [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    func showError(message: String) {
+        errorLabel.text = message
+        errorLabel.isHidden = false
+        borderImage.tintColor = .error
+    }
+}
+
+// allow only decimal numbers to be entered in the amount field
+extension TokenAmountField: UITextFieldDelegate {
+    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/Views/TokenAmountField.swift:50]: will that allow pasting negative numbers?

Comment (DmitryBespalov) [reply Multisig/UI/UI Library/Views/TokenAmountField.swift:50]: via copy-paste

File: Multisig/UI/UI Library/Views/TokenAmountField.swift [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        
+        let decimalSeparator = Locale.autoupdatingCurrent.decimalSeparator ?? "."
+        let inverseSet = CharacterSet(charactersIn:"0123456789").inverted
+        let components = string.components(separatedBy: inverseSet)
+        let filtered = components.joined(separator: "")
+        if filtered == string {
+            return true
+        } else {
+           if string.contains(decimalSeparator) {
+           let countdots = textField.text!.components(separatedBy: decimalSeparator).count - 1
```
Comment (DmitryBespalov) [inline Multisig/UI/UI Library/Views/TokenAmountField.swift]: please handle the optional  properly, this actually might crash here I think

File: Multisig/UI/UI Library/Views/TokenAmountField.swift (hunk 5 above)
Comment (elgatovital) [reply Multisig/UI/UI Library/Views/TokenAmountField.swift:50]: no. But we also don't want negative amounts, right?

Comment (DmitryBespalov) [reply Multisig/UI/UI Library/Views/TokenAmountField.swift:50]: right, just making sure that it should not accept negative numbers even if I copy-paste a negative number

**Review by MouazAlzahabi — APPROVED** [review]

File: Multisig/UI/UI Library/Views/TokenAmountField.swift (hunk 3 above)
Comment (elgatovital) [reply Multisig/UI/UI Library/Views/TokenAmountField.swift]: decimal separator used should adhere to locale according to https://github.com/gnosis/safe/wiki/How-to-format-amounts. 
Using Locale.current.decimalSeparator

Comment (DmitryBespalov) [reply Multisig/UI/UI Library/Views/TokenAmountField.swift]: it doesn't, there are combinations that even if locale is set to one value, the actual decimal separator that keyboard produces is something else

Comment (DmitryBespalov) [reply Multisig/UI/UI Library/Views/TokenAmountField.swift]: try simulating different locales + regions: US, German, UK, Japanese, Korean, French and see which decimal separator characters you're getting

Comment (elgatovital) [reply Multisig/UI/UI Library/Views/TokenAmountField.swift]: I have tried different locale/region combinations and each time I was getting the expected decimal separator. The same separator also was used in amounts displayed in the balances list.

Comment (elgatovital) [reply Multisig/UI/UI Library/Views/TokenAmountField.swift]: separator displayed in the decimalPad keyboard was adhering to locale in my tests.

**Review by DmitryBespalov — APPROVED** [review]

**Pull request merged**